Deprecate model registry stages in docs (#10480)
mlflow/mlflow: docs/source/model-registry.rst

    An MLflow Model is created from an experiment or run that is logged with one of the model flavor’s ``mlflow.<model_flavor>.log_model()`` methods. Once logged, this model can then be registered with the Model Registry.

Registered Model

>     An MLflow Model can be registered with the Model Registry. A registered model has a unique name, contains versions, version aliases, model lineage, and other metadata.

Reviewer: ```suggestion
    An MLflow Model can be registered with the Model Registry. A registered model has a unique name, contains versions, version aliases, version tags, model lineage, and other metadata.
```

>     An MLflow Model can be registered with the Model Registry. A registered model has a unique name, contains versions, version aliases, version tags, model lineage, and other metadata.

Reviewer 2: ```suggestion
    An MLflow Model can be registered with the Model Registry. A registered model has a unique name, contains versions, aliases, tags, and other metadata.
```

Minor but not sure it makes sense to mention "version tags" here since we (I think correctly) don't mention the other metadata fields on model versions here, e.g. model version descriptions

Model Version

>     Each registered model can have one or many versions. When a new model is added to the Model Registry, it is added as version 1. Each new model registered to the same model name increments the version number.

Reviewer 2: ```suggestion
    Each registered model can have one or many versions. When a new model is added to the Model Registry, it is added as version 1. Each new model registered to the same model name increments the version number. Model versions have tags, which can be useful for tracking attributes of the model version (e.g. `pre_deploy_checks: "PASSED"`)
```

Note that model alias assignments can be updated independently of your production code. If the ``champion`` alias in the snippet above is reassigned to a new model version in the Model Registry, the next execution of this snippet will automatically pick up the new model version. This allows you to decouple model deployments from your inference workloads.

**Fetch the latest model version in a specific stage**

To fetch a model version by stage, simply provide the model stage as part of the model URI, and it will fetch the most recent version of the model in that stage.

.. code-block:: python

    import mlflow.pyfunc

    model_name = "sk-learn-random-forest-reg-model"
    stage = "Staging"

    model = mlflow.pyfunc.load_model(model_uri=f"models:/{model_name}/{stage}")

    model.predict(data)

> 

Reviewer: We can add a small example code for representing how to set and use model alias here. It's ok to do it in follow-up PR.

New model deployment tools
--------------------------

Model stages were used to express the lifecycle of MLflow Models for productionization and deployment. Users transitioned model versions through four fixed stages (from **none**, to **staging**, to **production**, and then to **archived**) as they proposed, validated, deployed, and deprecated models for their ML use-cases. In doing so, model registry stages provided labeling and aliasing functionality for the model versions, by denoting the status of a model version in the UI and providing named references to model versions in the code (e.g. ``/Staging`` in the model URI). Model registry stages were also been used to denote the environment that the model is in.

To replace and improve upon stages, we elevated **model version tags**, introduced **model version aliases**, and enhanced the model registry UI to provide flexible and powerful ways to label and alias MLflow models in the Model Registry. We also introduced **environmental separation** to enable you to set up separate environments for your models and to express access controls for them.

**Model version tags**

> Now prominently displayed in the model registry UI, model version tags can be used to annotate model versions with their status. For example, you could apply a tag of key ``validation_status`` and value ``pending`` to a model version while it is being validated and then update the tag value to ``passed`` when it has passed smoke tests and performance tests.

Reviewer 2: ```suggestion
Model version tags can be used to annotate model versions with their status. For example, you could apply a tag of key ``validation_status`` and value ``pending`` to a model version while it is being validated and then update the tag value to ``passed`` when it has passed smoke tests and performance tests.
```

**Model version aliases**

> Model version aliases provide a flexible way to set named aliases on model versions. For example, setting a **champion** alias on a model version enables you to fetch this model version by that alias via the :meth:`~mlflow.client.MlflowClient.get_model_version_by_alias` client API or the model URI ``models:/<registered model name>@champion``. Aliases can be easily reassigned to new model versions via the UI and client API alike, thereby decoupling model deployment from the production system code. Unlike model registry stages, more than one alias can be applied to any given model version, creating powerful possibilities for model deployment.

Reviewer 2: ```suggestion
Model version aliases provide a flexible way to created named references for particular model versions, and are useful for identifying which model version(s) are deployed within an environment. For example, setting a **champion** alias on a model version enables you to fetch the model version by that alias via the :meth:`~mlflow.client.MlflowClient.get_model_version_by_alias` client API or the model URI ``models:/<registered model name>@champion``. Aliases can be reassigned to new model versions via the UI and client API. Unlike model registry stages, more than one alias can be applied to any given model version, creating powerful possibilities for model deployment.
```

Some suggested edits, I didn't quite get what "decoupling model deployment from the production system code" meant. Also maybe we can reorder this to something like "environment separation", "model version aliases", "model version tags" to focus on the top CUJs first?

Author: > "decoupling model deployment from the production system code" meant

By this, I meant that you can change which model is deployed just by updating the model version underlying the alias. You don't have to change your production code to point to a different model version number.

Author: > Also maybe we can reorder this to something like "environment separation", "model version aliases", "model version tags" to focus on the top CUJs first?

For this, I was thinking that it would be best to introduce new concepts last. After all, "environments" is a new concept we're bringing into MLflow. We should work up to it. Then, in the migration guidance section, we start off talking about environments.

Author: Made some more edits. I'm still leaning towards keeping the same order for new model deployment tools. The current ordering is consistent with order of functionality in the paragraph describing model registry stages, which I find works well.

> **[New] Set up separate environments for models**

Reviewer 2: ```suggestion
**Set up separate environments for models**
```

Should we remove the "new" block? Or have some issue etc to track removing it? We already mention "New deployment tools" above and this will become not-new in 2.10.0, so not sure it makes sense to keep here. Creating & ACLing separate registered models is also possible in older MLflow versions

Author: Makes sense!

> **[New] Environmental separation**

Reviewer 2: ```suggestion
**[New] Set up separate environments for models**
```

Or something? We could ask someone from docs team to review / suggest a more standard term here

In mature DevOps and MLOps workflows, organizations use separate environments (typically, dev, staging, and prod) with access controls to enable quick development without compromising stability in production. With :ref:`MLflow Authentication <auth>`, you can use registered models to express access-controlled environments for your MLflow models. For example, you can create registered models corresponding to each combination of environment and business problem (e.g. `prod.ml_team.revenue_forecasting`, `dev.ml_team.revenue_forecasting`) and configure permissions accordingly. Automate model retraining against your production registered models, or for simple model deployment use cases, use :meth:`~mlflow.client.MlflowClient.copy_model_version` to promote model versions across registered models.

Migrating models away from stages
---------------------------------

See the information below to learn how to use the new tools for your legacy Model Stage use-cases.

**Model environments**

> To set up separate environments and ACLs for your model versions, create separate registered models:

Reviewer 2: ```suggestion
To set up separate environments and permissions for your model versions, create separate registered models:
```
Minor, maybe replace "ACL" with "permissions" since seems our auth docs https://mlflow.org/docs/latest/auth/index.html use that terminology

* Given a base name for your model’s use-case, e.g. ``revenue_forecasting``, set up various registered models corresponding to your environments with different prefixes.
* For example, if you want three separate dev, staging, and production environments, you can set up ``dev.ml_team.revenue_forecasting``, ``staging.ml_team.revenue_forecasting``, and ``prod.ml_team.revenue_forecasting`` registered models.

> * Use :ref:`MLflow Authentication <auth>` to set up appropriate ACLs to these models.

Reviewer 2: ```suggestion
* Use :ref:`MLflow Authentication <auth>` to grant appropriate permissions on these models.
```

* For more mature production-grade setups, we recommend promoting your ML code (including model training code, inference code, and ML infrastructure as code) across environments. This eliminates the need to transition models across environments. Dev ML code is experimental and in a dev environment, hence targeting the dev registered model. Before merging developed ML code into your source code repository, your CI stages the code in a staging environment for integration testing (targeting the staging registered model). Post-merge, the ML code is deployed to production for automated retraining (targeting the prod registered model). Such setups enable safe and robust CI/CD of ML systems - including not just model training, but also feature engineering, model monitoring, and automated retraining.

**Model aliasing**

To specify (via named references) which model version to deploy to serve traffic within an environment (e.g. production), use **model aliases**:

> 1. Decide on an equivalent model alias for each model registry stage (e.g., **champion** for the **Production** stage)

Reviewer 2: ```suggestion
1. Decide on an equivalent model alias for each model registry stage (e.g., **champion** for the **Production** stage, to specify the model intended to serve the majority of traffic)
```
